### PR TITLE
feat(api): get document metadata

### PR DESCRIPTION
**Pull request by LetsMakeItTechnical**
## Describe your changes

This pull request updates the code for a router to handle HTTP GET requests to retrieve the properties/metadata of a single file on a case. The updated code includes Swagger annotations to document the API endpoint.

The router accepts two parameters in the URL path - id, which is the application ID, and guid, which is the GUID of the required document. The asyncHandler function wraps the getDocumentProperties function, which retrieves the document metadata and sends it in the HTTP response.

The Swagger annotations include tags, path, description, parameters, and response schema to provide clear documentation for the API endpoint. The response schema references the documentsPropertiesRequestBody definition, which describes the structure of the metadata for a single document.

The pull request also includes any necessary code changes to ensure that the router works as expected.

I have also added the @supercharge/promise-pool npm package for concurrent execution of asynchronous tasks. This package provides a clean and efficient way to execute multiple promises simultaneously, which can improve the performance and scalability of the application.

One of the key advantages of this package is that it allows you to control the concurrency level of your promises, which means you can specify how many promises you want to execute at once. This can be especially helpful when dealing with large sets of data or long-running tasks, as it allows you to optimise the use of system resources and prevent resource exhaustion.

Another benefit of using the @supercharge/promise-pool package is that it provides built-in error handling and retry functionality, which can save a lot of time and effort when dealing with asynchronous errors. The package automatically retries failed promises up to a specified number of times, and also provides options for handling errors and logging them in a consistent way.

In addition to these features, the @supercharge/promise-pool package also offers a clean and easy-to-use API for executing promises, making it a great choice.

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [ ] I have performed a self-review of my own code
- [ ] I have double checked this work does not include any hardcoded secrets or passwords
- [ ] I have made corresponding changes to the documentation
- [ ] I have provided details on how I have tested my code
- [ ] I have referenced the ticket number above
- [ ] I have provided a description of the ticket
- [ ] I have included unit tests to cover any testable code changes

### CODE REVIEW

File: apps/api/prisma/schema.d.ts [hunk 1]
```diff
@@ -187,23 +188,35 @@ export type CaseStatusNameType =
 	| 'Withdrawn';
 
 export interface DocumentDetails {
```
Comment (francescobrenci-AR) [inline apps/api/prisma/schema.d.ts:190]: doesn't this naming convention apply to all the other files as well?

File: apps/api/prisma/schema.d.ts [hunk 2]
```diff
@@ -188,24 +188,24 @@ export type CaseStatusNameType =
...
 	blobStoragePath: string | null;
-	blobStorageContainer: string | null | undefined;
-	from: string | null;
+	blobStorageContainer: string;
+	from: string;
 	receivedDate: number | null;
-	size: number | null;
-	type: string | null;
+	size: number;
+	filerType: string;
```
Comment (DavidJWhiteDev) [inline apps/api/prisma/schema.d.ts:200]: I thought this spelling got fixed?

File: apps/api/src/server/swagger.js [hunk 3]
```diff
@@ -180,6 +180,26 @@ const document = {
...
+			blobStorageContainer: 'document-service-uploads',
+			blobStoragePath: '/application/BC010001/1111-2222-3333/my doc.pdf',
+			from: 'joe blogs',
+			receivedDate: 1_677_585_578,
+			size: 0,
+			fileType: '',
+			redacted: false,
+			status: 'awaiting_upload',
+			description: '',
+			documentReferenceNumber: '',
```
Comment (DavidJWhiteDev) [inline apps/api/src/server/swagger.js:198]: should documentReferenceNumber be removed?  It looks like elsewhere you have replaced documentReferenceNumber with documentReference (which is also already in line 185 above).?

File: apps/api/prisma/schema.prisma [hunk 4]
```diff
@@ -3,8 +3,9 @@ generator client {
 }
 
 datasource db {
-  provider = "sqlserver"
-  url      = env("DATABASE_URL")
+  provider          = "sqlserver"
+  url               = env("DATABASE_URL")
+  shadowDatabaseUrl = env("sqlserver://127.0.0.1:1433;database=pinss_development;user=sa;password=<YourStrong@Passw0rd>;trustServerCertificate=true")
```
[the remark this answers is not in the document]
Comment (LetsMakeItTechnical) [reply apps/api/prisma/schema.prisma:8]: accident

**Review by DavidJWhiteDev — APPROVED** [review]

File: apps/api/src/server/utils/mapping/map-document-details.js [hunk 5]
```diff
@@ -1,36 +1,104 @@
...
-		redacted: documentDetails.redacted,
-		status: documentDetails.status,
-		description: '',
-		documentReferenceNumber: '',
-		version: 1,
-		agent: '',
-		caseStage: '',
-		webFilter: '',
-		documentType: ''
+		documentGuid: documentMetadata.documentGuid,
```
Comment (lutamKC) [inline apps/api/src/server/utils/mapping/map-document-details.js:15]: hey, I tried and seems like this line creates an error.
Spefically the API _/applications/{caseId}/folders/{folderId}/documents_ returns this error:

```
{
  "errors": "Cannot read properties of null (reading 'documentGuid')"
}
```

do you know what could cause it?

**Review by lutamKC — CHANGES_REQUESTED** [review]

File: apps/api/prisma/schema.d.ts [hunk 6]
```diff
@@ -187,23 +189,45 @@ export type CaseStatusNameType =
 	| 'Withdrawn';
 
 export interface DocumentDetails {
-	guid: string;
+	documentId: string | null;
+	sourceSystem: string;
+	documentGuid: string | null;
+	redacted: boolean;
```
Comment (francescobrenci-AR) [inline apps/api/prisma/schema.d.ts:195]: this is duplicated

Comment (francescobrenci-AR) [reply apps/api/prisma/schema.d.ts:195]: also need some updates form this morning's chat in Teams - updated list should be available soon https://pins-ds.atlassian.net/wiki/spaces/BO/pages/1064730651/Document+Metadata+-+DRAFT

Comment (LetsMakeItTechnical) [reply apps/api/prisma/schema.d.ts:195]: @francescobrenci-AR whats duplicate

File: apps/api/src/server/applications/application/documents/document.service.js [hunk 7]
```diff
@@ -126,45 +146,20 @@ export const obtainURLsForDocuments = async (documentsToUpload, caseId) => {
...
-		}
-	];
-
-	return documentMetadata.find((data) => data.id === id);
+export const upsertDocumentMetadataAndReturnDetails = async (
+	documentGuid,
+	documentMetadataBody
+) => {
+	const documentMetadata = await documentMetadataRepository.upsert({
+		...documentMetadataBody,
```
Comment (chrisdalykinandcarta) [inline apps/api/src/server/applications/application/documents/document.service.js:160]: I think we might be exposing a [mass assignment](https://cheatsheetseries.owasp.org/cheatsheets/Mass_Assignment_Cheat_Sheet.html) vulnerability here - looking at the joi schema, we allow the request to set things like the virus check status which I don't think the end user should ever be able to do.

This joi schema should describe the request model that the API allows (I guess this should just be details which the end user can modify).

Comment (LetsMakeItTechnical) [reply apps/api/src/server/applications/application/documents/document.service.js:178]: @chrisdalykinandcarta I did think that but I think we need to discuss a team what fields should be mutable to the end user and what can’t

Comment (LetsMakeItTechnical) [reply apps/api/src/server/applications/application/documents/document.service.js:178]: @chrisdalykinandcarta I also think at later point we need to add role management

**Review by francescobrenci-AR — DISMISSED** [review]

**Review by lutamKC — DISMISSED** [review]

**Review by francescobrenci-AR — APPROVED** [review]

**Review by francescobrenci-AR — DISMISSED** [review]

**Review by lutamKC — APPROVED** [review]

**Pull request closed without merging**